advanced/wallets/react-wallet-v2/src/lib/smart-accounts/SmartAccountLib.ts

```diff
@@ -0,0 +1,174 @@
+import {
+  BundlerActions,
+  BundlerClient,
+  SmartAccountClient,
+  SmartAccountClientConfig,
+  bundlerActions,
+  createSmartAccountClient
+} from 'permissionless'
+import { PimlicoBundlerActions, pimlicoBundlerActions } from 'permissionless/actions/pimlico'
+import { PIMLICO_NETWORK_NAMES, UrlConfig, publicRPCUrl } from '@/utils/SmartAccountUtil'
+import { Chain } from '@/consts/smartAccounts'
+
+type SmartAccountLibOptions = {
+  privateKey: string
+  chain: Chain
+  sponsored?: boolean
+}
+
+export abstract class SmartAccountLib implements EIP155Wallet {
+  // Options
+  public chain: Chain
+  public sponsored: boolean = true
+
+  // Signer
+  protected signer: PrivateKeyAccount
+  #signerPrivateKey: string
+
+  // Clients
+  protected publicClient: PublicClient
+  protected paymasterClient: PimlicoPaymasterClient
+  protected bundlerClient: BundlerClient & BundlerActions & PimlicoBundlerActions
+  protected client: (SmartAccountClient & PimlicoBundlerActions) | undefined
+
+  // Transport
+  protected bundlerUrl: HttpTransport
+  protected paymasterUrl: HttpTransport
+
+  // Utility
+  public type: string
+  public initialized = false
+
+  public constructor({ privateKey, chain, sponsored = false }: SmartAccountLibOptions) {
+    const apiKey = process.env.NEXT_PUBLIC_PIMLICO_KEY
+    const paymasterUrl = ({ chain }: UrlConfig) =>
+      `https://api.pimlico.io/v2/${PIMLICO_NETWORK_NAMES[chain.name]}/rpc?apikey=${apiKey}`
+    const bundlerUrl = ({ chain }: UrlConfig) =>
+      `https://api.pimlico.io/v1/${PIMLICO_NETWORK_NAMES[chain.name]}/rpc?apikey=${apiKey}`
+
+    this.chain = chain
+    this.sponsored = sponsored
+    this.#signerPrivateKey = privateKey
+    this.signer = privateKeyToAccount(privateKey as Hex)
+
+    this.bundlerUrl = http(bundlerUrl({ chain: this.chain }))
+    this.paymasterUrl = http(paymasterUrl({ chain: this.chain }))
+
+    this.publicClient = createPublicClient({
+      transport: http(publicRPCUrl({ chain: this.chain }))
+    }).extend(bundlerActions)
+
+    this.paymasterClient = createPimlicoPaymasterClient({
+      transport: this.paymasterUrl
+    })
+
+    this.bundlerClient = createClient({
+      transport: this.bundlerUrl,
+      chain: this.chain
+    })
+      .extend(bundlerActions)
+      .extend(pimlicoBundlerActions)
+
+    const name = this.constructor.name.replace('SmartAccountLib', '')
+    this.type = name && name.length > 0 ? name : 'Unknown'
+  }
+
+  abstract getClientConfig(): Promise<SmartAccountClientConfig>
+
+  async init() {
+    const config = await this.getClientConfig()
+    this.client = createSmartAccountClient(config).extend(pimlicoBundlerActions)
+    console.log('Smart account initialized', {
+      address: this.client.account?.address,
+      chain: this.chain.name,
+      type: this.type
+    })
+    this.initialized = true
+  }
+
+  getMnemonic(): string {
+    throw new Error('Method not implemented.')
+  }
+  getPrivateKey(): string {
```

Comment:
what happens if this is instanced with a signer?

Reply:
What do you mean by signer?

Reply:
instead of using a privatekey you can instantiate the smart account client using a signer (e.g. the magic signer)

Reply:
Currently we only instantiate this with a private key. Impossible to do so in a different manner. If there is a need, we can extend this to support mnemonics as well. Left that method because I introduced `EIP155Wallet` interface in the last PR as well